driver-core/src/main/com/mongodb/internal/connection/BaseCluster.java

```diff
@@ -462,7 +457,7 @@ private void notifyWaitQueueHandler(final ServerSelectionRequest request) {
             waitQueue.add(request);
 
             if (waitQueueHandler == null) {
-                waitQueueHandler = new Thread(new WaitQueueHandler(), "cluster-" + clusterId.getValue());
+                waitQueueHandler = new Thread(new WaitQueueHandler(operationContext), "cluster-" + clusterId.getValue());
```

Comment:
This looks wrong - as the waitQueueHandler only gets the first operationContext?

Reply:
(I rebased the original 3 commits, after merging master. This removed getDescription methods.)

It looks like there is a bug in our server selection code. On master, the `waitForSrv` method initiates a server selection timeout. However, this method is invoked by getServer, which is invoked from BaseCluster (handleServerSelectionRequest...), and ultimately this call chain comes to where a new ServerSelectionRequest is created. This creation has its own server selection start time (assigned to a field). It is this original server selection timeout that should be used by `waitForSrv`, but instead, it creates its own, which could be significantly later than the original start.

While making these changes, I also removed the StartTime from timeout error messages (instead of threading through that startTime). This part of the message will be removed for standardized logging.

One moral is to be careful if one sees an OperationContext being used to create a timeout, but then being passed to downstream methods.

Reply:
The situation above, with double timeouts, never arises because the affected getServer in LoadBalancedCluster is never invoked by BaseCluster (LBC not a subclass of BC).